editor/filesystem_dock.cpp

```diff
@@ -2705,6 +2741,53 @@ void FileSystemDock::_get_drag_target_folder(String &target, bool &target_favori
 	}
 }
 
+void FileSystemDock::_folder_color_index_pressed(int p_index) {
+	Dictionary settings = ProjectSettings::get_singleton()->get_setting("file_customization/folder_colors");
+	Variant chosen_color_name = folder_colors_menu->get_item_metadata(p_index);
+	Vector<String> selected;
+
+	// Get all selected folders based on whether the files panel or tree panel is currently focused.
+	if (files->has_focus()) {
+		Vector<int> files_selected_ids = files->get_selected_items();
+		for (int i = 0; i < files_selected_ids.size(); i++) {
+			selected.push_back(files->get_item_metadata(files_selected_ids[i]));
+		}
+	} else {
+		TreeItem *tree_selected = tree->get_root();
+		tree_selected = tree->get_next_selected(tree_selected);
+		while (tree_selected) {
+			selected.push_back(tree_selected->get_metadata(0));
+			tree_selected = tree->get_next_selected(tree_selected);
+		}
+	}
+
+	// Update project settings with new folder colors.
+	for (int i = 0; i < selected.size(); i++) {
+		String fpath = selected[i];
+
+		if (chosen_color_name) {
+			settings[fpath] = chosen_color_name;
+		} else {
+			settings.erase(fpath);
+		}
+	}
+
+	ProjectSettings::get_singleton()->set_setting("file_customization/folder_colors", settings);
```

Comment:
Dictionaries are passed by reference, so this shouldn't be necessary.

Reply:
Not necessarily true here, as if the dictionary is *empty* which it is when starting the settings will be returned as a default value when fetched and not properly tracked

Reply:
Right, but if the setting existed, this would do a redundant set.

Reply:
Sure, but the setting doesn't exist for any new user, and won't ever unless they already had it, which they won't